docs/design/federated-services.md

```diff
@@ -0,0 +1,318 @@
+# Kubernetes Cluster Federation (a.k.a. "Ubernetes")
+
+## _Cross-cluster Load Balancing and Service Discovery  
+### Requirements and System Design_
+
+### by Quinton Hoole  
+### Dec 3 2015
+
+## Requirements
+
+### Discovery, Load-balancing and Failover
+
+1. **Internal discovery and connection**: Pods/containers (running in a Kubernetes cluster) must be able to easily discover and connect to endpoints for Kubernetes services on which they depend in a consistent way, irrespective of whether those services exist in a different kubernetes cluster within the same cluster federation.  Hence-forth referred to as "cluster-internal clients", or simply "internal clients".
+1. **External discovery and connection**: External clients (running outside a Kubernetes cluster) must be able to discover and connect to endpoints for Kubernetes services on which they depend.  
+   1. **External clients predominantly speak HTTP(S)**: External clients are most often, but not always, web browsers, or at least speak HTTP(S) (TBD list notable exceptions.  DNS servers?  SIP servers?  Database Servers?)
+1. **Find the "best" endpoint:** Upon initial discovery and connection, both internal and external clients should ideally find "the best" endpoint if multiple eligible endpoints exist.  "Best" in this context implies the closest (by network topology) endpoint that is both operational (as defined by some positive health check) and not overloaded (by some published load metric).  For example:
```

Comment:
@lavalamp By "this" are you referring to "not overloaded" specifically, or "best endpoint" generally. The former we can definitely live without, although in practise it is extremely useful in reducing human operational toil and making capacity planning way, way simpler.  So I think we'd definitely want to add it over time.  The latter ("best endpoint") we can also technically do without, although it has pretty major latency (and hence performance) and reliability implications, so it would be a higher priority than "not overloaded".